### ISSUE TITLE
Wizard form not updating form state with InitialValues on subsequent pages

### ISSUE THREAD

**Issue by MitchWeiss** [erikras/redux-form#679]
I'm building a wizard form using [http://erikras.github.io/redux-form/#/examples/wizard](url) as a guide

I have noticed that the first page of the form will immediately put the form fields and any initial values into the form state, but any fields on subsequent pages will not be added until they are touched. This means that if the user never touches the field and leaves the default value, the field will return 'undefined' on submission. 

I believe this is a bug but please let me know if I am configuring the form incorrectly. I have tried to make the simplest example possible to highlight the issue:

```
// multiform.js

import React, { Component } from 'react';
import FirstForm from './first_form';
import SecondForm from './second_form';

export default class MultiForm extends Component {
  constructor(props) {
    super(props);
    this.state = {step: 1};
    this.nextStep = this.nextStep.bind(this);
    this.previousStep = this.previousStep.bind(this);
  }

  nextStep() {
    this.setState({step: this.state.step + 1});
  }

  previousStep() {
    this.setState({step: this.state.step - 1});
  }

  handleSubmit(data) {
    console.log(data);
  }

  render() {
    const {page} = this.state;
    return (
      <div>
        {this.state.step === 1 && <FirstForm onSubmit={this.nextStep}/>}
        {this.state.step === 2 && <SecondForm onSubmit={this.handleSubmit} previousStep={this.previousStep}/>}
      </div>
    )
  }
}
```

```
// first_form.js

import React, { Component } from 'react';
import { reduxForm } from 'redux-form';

class FirstForm extends Component {
  render() {
      const { handleSubmit } = this.props;
      const { fields: { testIn, testSelect, testRadio } } = this.props;
      return (
        <form onSubmit={handleSubmit}>
          <div>
            <input type="text" {...testIn} />
          </div>

          <div>
            <select type="select" {...testSelect} value={testSelect.value}>
              <option value="a">AAA</option>
              <option value="b">BBB</option>
            </select>
          </div>

          <div>
            <input type="radio" {...testRadio} value="x" checked={testRadio.value === 'x'}/> XXX
            <input type="radio" {...testRadio} value="y" checked={testRadio.value === 'y'}/> YYY
          </div>

          <div>
            <button type="button" type="submit">Next</button>
          </div>
        </form>
      );
  }
}

FirstForm = reduxForm({
  form: 'anotherForm',
  fields: ['testIn', 'testSelect', 'testRadio'],
  initialValues: {testIn: "abc", testSelect: "b", testRadio:"y"},
  destroyOnUnmount: false
})(FirstForm);

export default FirstForm;
```

```
// second_form.js

import React, { Component } from 'react';
import { reduxForm } from 'redux-form';

class SecondForm extends Component {
  render() {
      const { handleSubmit, previousStep } = this.props;
      const { fields: { testIn2, testSelect2, testRadio2 } } = this.props;

      return (
        <form onSubmit={handleSubmit}>
          <div>
            <input type="text" {...testIn2} />
          </div>

          <div>
            <select type="select" {...testSelect2} value={testSelect2.value}>
              <option value="c">CCC</option>
              <option value="d">DDD</option>
            </select>
          </div>

          <div>
            <input type="radio" {...testRadio2} value="i" checked={testRadio2.value === 'i'}/> III
            <input type="radio" {...testRadio2} value="j" checked={testRadio2.value === 'j'}/> JJJ
          </div>

          <div>
          <button type="button" onClick={previousStep}>Previous</button>
            <button type="button" type="submit">Submit</button>
          </div>
        </form>
      );
  }
}

SecondForm = reduxForm({
  form: 'anotherForm',
  fields: ['testIn2', 'testSelect2', 'testRadio2', 'testIn', 'testSelect', 'testRadio'],
  initialValues: {testIn2: "def", testSelect2: "d", testRadio2:"j"},
  destroyOnUnmount: false
})(SecondForm);

export default SecondForm;
```

Submitting the form without touching any field, only clicking the next and submit buttons outputs the following to the console:

```
testIn: "abc"
testIn2: undefined
testRadio: "y"
testRadio2: undefined
testSelect: "b"
testSelect2: undefined
```

**Comment by ptim** [comment]
Great report.. 

I'm also finding that on page 2+ of my wizard form: prefilled values with a `required` validator are not touched on submit, so I need to go manually touch each field before it will validate... will update with details as I find them.

**Comment by MitchWeiss** [comment]
Until this is properly looked at by Erik, my workaround has been to include all the fields I'm providing an initialValue to in the first page. This seems to have worked well and has also fixed the issue I was experiencing where radio buttons which were provided an initialValue were not being selected on load (https://github.com/erikras/redux-form/issues/124#issuecomment-187039217)

```
FirstForm = reduxForm({
  form: 'anotherForm',
  fields: ['testIn', 'testSelect', 'testRadio', 'testIn2', 'testSelect2', 'testRadio2'],
  initialValues: {testIn: "abc", testSelect: "b", testRadio:"y", testIn2: "def", testSelect2: "d", testRadio2:"j"},
  destroyOnUnmount: false
})(FirstForm);
```

**Comment by ptim** [comment]
Thanks! awesome works for me :)

**Comment by erikras** [comment]
@MitchWeiss Yes, that is the solution. At the moment, `initialize()` needs the entire `fields` array to function properly. I hope to remedy this in future versions, but at the moment, it's a fact of life.

Great reporting! :+1:

**Comment by MitchWeiss** [comment]
Great, thanks for the confirmation @erikras. May I suggest putting a quick note about this into the wizard form example? I know it definitely had me tearing my hair out trying to work out why it wasn't working

**Comment by lauterry** [comment]
Hello

It seems that there is a problem with @MitchWeiss solution.

When moving to step 2 of the wizard, the action ```redux-form/TOUCH``` is triggered, making all fields of step 2 to be touched. The result is that errors messages of step 2 fields are displayed since the validation of the fields are run. (correct me if i'm wrong).

Best regards

**Comment by ooflorent** [comment]
This is a usage question, and we encourage you to ask it on StackOverflow instead. We try to use the issue tracker for bug reports and feature requests, as well as some discussions, but this appears to be a very localized question.

This is the only way to ensure the same question doesn’t get asked many times in a row, and that folks who have the same question in the future will be able to find it.

I’m closing for these reasons. Thanks to everyone who gave his feedback.

**Comment by lauterry** [comment]
Hello

The problem I've found is a typical use case of a wizard form.

Let me describe it more precisely : 

I've 2 forms with the same form name. Let's say all fields are required and can display an error thank to the following code : 

```{ firstname.touched && firstname.error ? <div>{ firstname.error }</div> : false }```

Since I need to pass initialValues to some fields of form 2 (**not all fields**), the only way to do that actually is the workaround described by @MitchWeiss [here](https://github.com/erikras/redux-form/issues/679#issuecomment-188622841).

When I submit the first form, all fields of form 1 and form 2 are touched because of [this line of code](https://github.com/erikras/redux-form/blob/master/src/handleSubmit.js#L10)

The result is that the fields of form 2 which don't have initialValues will display an error because they are **empty AND touched**.

I share my workaround here which consist on untouching all fields of form 2. To do that, I put the following code in the form 2 component : 

```
componentWillMount() {
        this.props.untouchAll();
}
```

Hope for a better solution will be found for futures versions but please keep in mind my use case :) 

**By the way, Thanks for this great lib 👍 👍 **

Best regards

**Comment by MitchWeiss** [comment]
Hey @lauterry 

Don't list the fields for form 2 in form 1 unless they have an initial value, that way they won't get marked as touched when you move from form 1 to form 2.

**Comment by lauterry** [comment]
Hello @MitchWeiss 

Thank for you reply which lead me to think that I have to **dynamically** add the fields of form 2 in form 1 depending they have initial values or not.

In my case, it depends whether they can be initiated from locale storage or not.

Maybe one should not mark this issue as closed until a better solution is found in the future to handle InitialValues across wizard steps.

Best regards

**Comment by MitchWeiss** [comment]
I haven't had a chance to try out v6 but this issue seems to be solved in that version #850

**Comment by lauterry** [comment]
Oh thanks to point this out !
